crates/sui-json-rpc/src/indexer_api.rs

```diff
@@ -371,26 +380,60 @@ impl<R: ReadApiServer> IndexerApiServer for IndexerApi<R> {
     #[instrument(skip(self))]
     async fn resolve_name_service_address(&self, name: String) -> RpcResult<Option<SuiAddress>> {
         with_tracing!(async move {
-            let domain = Domain::from_str(&name).map_err(|e| {
-                Error::UnexpectedError(format!(
-                    "Failed to parse NameService Domain with error: {:?}",
-                    e
-                ))
-            })?;
-
+            // prepare the requested domain's field id.
+            let domain = name.parse::<Domain>().map_err(Error::from)?;
             let record_id = self.name_service_config.record_field_id(&domain);
 
-            let field_record_object = match self.state.get_object(&record_id).await? {
-                Some(o) => o,
-                None => return Ok(None),
+            // prepare the parent's field id.
+            let parent_domain = domain.parent();
+            let parent_record_id = self.name_service_config.record_field_id(&parent_domain);
+
+            let current_timestamp_ms = self.get_latest_checkpoint_timestamp_ms()?;
+
+            // Do these two reads in parallel.
+            let mut requests = vec![self.state.get_object(&record_id)];
+
+            // Also add the parent in the DB reads if the requested domain is a subdomain.
+            if domain.is_subdomain() {
+                requests.push(self.state.get_object(&parent_record_id));
+            }
+
+            // Couldn't find a `multi_get_object` for this crate (looks like it uses a k,v db)
+            // Always fetching both parent + child at the same time (even for node subdomains),
+            // to avoid sequential db reads.
```

Comment:
Will add some extra context on the title, but it's indeed based off that! 

A leaf name's expiration is derived from the parent.. and we can't know whether a name we're querying is leaf beforehand (so we have to query both).

The reason I am building it this way, is that I want to do a single `multi_get_objects` on the new systems (indexer_v2 & graphql) to avoid the 2 db round-trips!